cloud_storage: make cluster default read/write properties "sticky"

jcsp opened the pull request:
## Cover letter

Instead of treating remote.read and remote.write as properties that can be overridden by topics or can be null and thereby get cluster defaults (including updating when those defaults change), we will now always store an explicit value for the properties on the topic, and the cluster defaults will simply control what that explicit value defaults to at time of topic creation (or at time of upgrade, for clusters coming from <= 22.2)

This is based on https://github.com/redpanda-data/redpanda/pull/6876

Fixes https://github.com/redpanda-data/redpanda/issues/6917

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

See release notes

## Release notes

### Improvements

* The properties `cloud_storage_enable_remote_[read|write]` are now applied to topics at creation time, and if they subsequently change, then existing topics' properties do not change.  To modify the tiered storage mode of existing topics, you may set the `redpanda.remote.[read|write]` properties on the topic.

VladLazar reviewed (commented):
Looking good. I think we should update `disk_log_impl::override_retention_config` to remove the usage `cloud_storage_enable_remote_write` and use whatever is in the `ntp_config`. Might be worth waiting for https://github.com/redpanda-data/redpanda/pull/6951 to merge though.

jcsp commented:
>I think we should update disk_log_impl::override_retention_config to remove the usage cloud_storage_enable_remote_write and use whatever is in the ntp_config

Agreed, added a commit that does that and rebased on dev since #6951 merged

jcsp commented:
Oops, the last force push picked up some unrelated commits, re-pushed.

jcsp commented:
Force push: update `test_bytes_eviction_overrides` for the way we no longer take topic SI enablement at the storage layer from the global config, this PR raced with the one that added it

VladLazar commented:
Code looks fine, but there's two tests failing:
* `ShadowIndexingRetentionTesttest_shadow_indexing_default_local_retention`: The logic for computing whether a deletion is expected needs updating to reflect the new behaviour:
`expect_deletion = cluster_remote_write if topic_remote_write == "-1" else topic_remote_write == "true"`
* `UpgradeFromPriorFeatureVersionCloudStorageTest.test_rolling_upgrade` also fails, but I'm not sure why

jcsp commented:
Pushed:
- Updated retention_policy_test for the way that enabling the cluster default property no longer forces SI on for every topic
- Updated archival_metadata_stm logic for `max_collectible_offset`: this was behind the `test_rolling_upgrade` failure: it was a legit bug, where on a partly upgraded system we were prefix-truncating the local log ahead of the uploaded location (more in commit message)

jcsp commented:
Push:
- One more commit to do the same upgrade handling trick in cluster::partition that I just did in archival state machine.  Basically these are the places where our interpretation of enabled-ness changed, and we need to interpret ntp configs in the old style until upgrades are complete.

The failure for this last commit is in test_rolling_upgrade but not deterministic, reproduced when running lots of test runs in parallel on my workstation.

dotnwat reviewed (commented):
code lgtm

VladLazar reviewed (approved):
Looks good. Nice catch on applying the old behaviour during upgrades.
